File: sample-objc/ARGEARSample/UI/UIComponent/Main/Bottom/MainBottomFunctionView.m
//
//  MainBottomFunctionView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "MainBottomFunctionView.h"
#import "UIImage+Color.h"

@implementation MainBottomFunctionView

#define kMainBottomFunctionBeautyButtonName @"icBeauty"
#define kMainBottomFunctionFilterButtonName @"icFilter"
#define kMainBottomFunctionStickerButtonName @"icSticker"
#define kMainBottomFunctionBulgeButtonName @"icBulge"

#define kMainBottomFunctionRecordTimeLabelRatioFullColor [UIColor whiteColor]
#define kMainBottomFunctionRecordTimeLabelColor [UIColor colorWithRed:97.f/255.f green:97.f/255.f blue:97.f/255.f alpha:1.f]
#define kMainBottomFunctionRecordTimeLabelShadowColor [UIColor colorWithRed:97.f/255.f green:97.f/255.f blue:97.f/255.f alpha:1.f]

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (void)awakeFromNib {
    [super awakeFromNib];
    
    [self addObservers];
    
    [self setButtons:[self tag]];
}

- (void)dealloc {
    [self removeObservers];
}

- (void)addObservers {
    [_modeSelectView addObserver:_shutterView forKeyPath:@"mode" options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld context:nil];
}

- (void)removeObservers {
    [_modeSelectView removeObserver:_shutterView forKeyPath:@"mode"];
}

- (void)setRecordTime:(CGFloat)time {
    int min = time / 60.f;
    int sec = time - min * 60.f;
    
    [_recordTimeLabel setText:[NSString stringWithFormat:@"%02d:%02d", min, sec]];
}

- (IBAction)beautyButtonAction:(UIButton *)sender {
    
    [self openBottomFunctions];
    
    [_beautyView open];
}

- (IBAction)filterButtonAction:(UIButton *)sender {
    
    [self openBottomFunctions];
    
    [_filterView open];
}

- (IBAction)stickerButtonAction:(UIButton *)sender {
    
    [self openBottomFunctions];
    
    [_stickerView open];
}

- (IBAction)bulgeButtonAction:(UIButton *)sender {
    [self openBottomFunctions];
    
    [_bulgeView open];
}

- (IBAction)shutterButtonAction:(UIButton *)sender {
    
    switch ([_modeSelectView mode]) {
        case ARGMediaModePhoto:
            if (_delegate && [_delegate respondsToSelector:@selector(takePictureAction:)]) {
                [_delegate takePictureAction:sender];
            }
            break;
        case ARGMediaModeVideo:
            [_recordTimeLabel setHidden:[sender tag]];
            [self setModeSelectViewDisable:[sender tag]];

            if (_delegate && [_delegate respondsToSelector:@selector(recordVideoAction:)]) {
                [_delegate recordVideoAction:sender];
            }
            break;
        default:
            break;
    }
}

- (void)openBottomFunctions {
    [_touchCancelView setHidden:NO];
    
    [self setHidden:YES];
}

- (void)closeBottomFunctions {
    [_touchCancelView setHidden:YES];
    
    [_beautyView close];
    [_filterView close];
    [_stickerView close];
    [_bulgeView close];
    
    [self setHidden:NO];
}

- (void)setModeSelectViewDisable:(BOOL)disable {
    [_modeSelectView setUserInteractionEnabled:disable];
}

- (IBAction)handleTapFrom:(UITapGestureRecognizer*)tap {
    if ([tap state] == UIGestureRecognizerStateEnded) {
        
        [self closeBottomFunctions];
    }
}

- (void)setButtons:(ARGMediaRatio)ratio {
    
    NSMutableString *beautyButtonImageString = [NSMutableString stringWithString:kMainBottomFunctionBeautyButtonName];
    NSMutableString *filterButtonImageString = [NSMutableString stringWithString:kMainBottomFunctionFilterButtonName];
    NSMutableString *stickerButtonImageString = [NSMutableString stringWithString:kMainBottomFunctionStickerButtonName];
    NSMutableString *bulgeButtonImageString = [NSMutableString stringWithString:kMainBottomFunctionBulgeButtonName];
    
    if (ratio == ARGMediaRatio_16x9) {
        [beautyButtonImageString appendString:@"White"];
        [filterButtonImageString appendString:@"White"];
        [stickerButtonImageString appendString:@"White"];
        [bulgeButtonImageString appendString:@"White"];
    } else {
        [beautyButtonImageString appendString:@"Black"];
        [filterButtonImageString appendString:@"Black"];
        [stickerButtonImageString appendString:@"Black"];
        [bulgeButtonImageString appendString:@"Black"];
    }
    
    [beautyButtonImageString appendString:[NSString stringWithFormat:@"_%@", [@"" getLanguageCode]]];
    [filterButtonImageString appendString:[NSString stringWithFormat:@"_%@", [@"" getLanguageCode]]];
    [stickerButtonImageString appendString:[NSString stringWithFormat:@"_%@", [@"" getLanguageCode]]];
    [bulgeButtonImageString appendString:[NSString stringWithFormat:@"_%@", [@"" getLanguageCode]]];
    
    [_beautyButton setImage:[UIImage imageNamed:[NSString stringWithString:beautyButtonImageString]] forState:UIControlStateNormal];
    [_beautyButton setImage:[[UIImage imageNamed:[NSString stringWithString:beautyButtonImageString]] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];

    [_filterButton setImage:[UIImage imageNamed:[NSString stringWithString:filterButtonImageString]] forState:UIControlStateNormal];
    [_filterButton setImage:[[UIImage imageNamed:[NSString stringWithString:filterButtonImageString]] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];

    [_stickerButton setImage:[UIImage imageNamed:[NSString stringWithString:stickerButtonImageString]] forState:UIControlStateNormal];
    [_stickerButton setImage:[[UIImage imageNamed:[NSString stringWithString:stickerButtonImageString]] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];

    [_bulgeButton setImage:[UIImage imageNamed:[NSString stringWithString:bulgeButtonImageString]] forState:UIControlStateNormal];
    [_bulgeButton setImage:[[UIImage imageNamed:[NSString stringWithString:bulgeButtonImageString]] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];
}

- (void)setRecordTimeLabelColor:(ARGMediaRatio)ratio {
    
    UIColor *textColor = [UIColor clearColor];
    UIColor *shadowColor = [UIColor clearColor];

    if (ratio == ARGMediaRatio_16x9) {
        
        textColor = kMainBottomFunctionRecordTimeLabelRatioFullColor;
        shadowColor = kMainBottomFunctionRecordTimeLabelShadowColor;
    } else {
        
        textColor = kMainBottomFunctionRecordTimeLabelColor;
    }
    
    [_recordTimeLabel setTextColor:textColor];
    
    [[_recordTimeLabel layer] setShadowOffset:CGSizeZero];
    [[_recordTimeLabel layer] setShadowRadius:1.f];
    [[_recordTimeLabel layer] setShadowOpacity:0.5f];
    [[_recordTimeLabel layer] setMasksToBounds:NO];
    [[_recordTimeLabel layer] setShouldRasterize:YES];
    [[_recordTimeLabel layer] setShadowColor:[shadowColor CGColor]];
}

- (void)changeRatio:(ARGMediaRatio)ratio {
    
    [self setButtons:ratio];
    [self setRecordTimeLabelColor:ratio];
    
    [_modeSelectView setRatio:ratio];
    [_shutterView setRatio:ratio];
    
    [_beautyView setRatio:ratio];
    [_filterView setRatio:ratio];
    [_stickerView setRatio:ratio];
    [_bulgeView setRatio:ratio];
}

# pragma mark - Observers
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    
    if ([keyPath isEqualToString:@"currentRatio"]) {
        ARGMediaRatio ratio = [[change objectForKey:@"new"] integerValue];
        
        [self changeRatio:ratio];
    }
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/Main/Bottom/MainBottomFunctionView.swift
//
//  MainBottomFunctionView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/30.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit
import ARGear

enum MainBottomButtonType : Int {
    case Beauty = 0
    case Filter = 1
    case Content = 2
    case Bulge = 3
}

protocol MainBottomFunctionDelegate: class {
    func photoButtonAction(_ button: UIButton)
    func videoButtonAction(_ button: UIButton)
}

class MainBottomFunctionView: UIView {
    
    let kMainBottomFunctionBeautyButtonName = "icBeauty"
    let kMainBottomFunctionFilterButtonName = "icFilter"
    let kMainBottomFunctionContentButtonName = "icSticker"
    let kMainBottomFunctionBulgeButtonName = "icBulge"

    let kMainBottomFunctionRecordTimeLabelRatioFullColor = UIColor.white
    let kMainBottomFunctionRecordTimeLabelColor = UIColor(red: 97.0/255.0, green: 97.0/255.0, blue: 97.0/255.0, alpha: 1.0)
    let kMainBottomFunctionRecordTimeLabelShadowColor = UIColor(red: 97.0/255.0, green: 97.0/255.0, blue: 97.0/255.0, alpha: 1.0)

    weak var delegate: MainBottomFunctionDelegate?

    @IBOutlet weak var touchCancelView: UIView!

    @IBOutlet weak var beautyButton: UIButton!
    @IBOutlet weak var filterButton: UIButton!
    @IBOutlet weak var contentButton: UIButton!
    @IBOutlet weak var bulgeButton: UIButton!

    @IBOutlet weak var modeSelectView: ModeSelectView!

    @IBOutlet weak var shutterButton: UIButton!
    @IBOutlet weak var shutterView: ShutterView!
    
    @IBOutlet weak var beautyView: BeautyView!
    @IBOutlet weak var filterView: FilterView!
    @IBOutlet weak var contentView: ContentView!
    @IBOutlet weak var bulgeView: BulgeView!
    
    @IBOutlet weak var recordTimeLabel: UILabel!
    
    private var argObservers = [NSKeyValueObservation]()

    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.setRatio(ARGMediaRatio(rawValue: self.tag) ?? ._4x3)
        
        self.addObservers()
    }
    
    deinit {
        self.removeObservers()
    }
    
    func addObservers() {
        self.argObservers.append(
            // mode change observer
            self.modeSelectView.observe(\.mode, options: [.new]) { [weak self] obj, _ in
                guard let self = self else { return }

                self.shutterView.setShutterMode(obj.mode)
            }
        )
    }
    
    func removeObservers() {
        self.argObservers.removeAll()
    }
    
    func setRatio(_ ratio: ARGMediaRatio) {
        
        self.setButtonsImage(ratio: ratio)
        self.setRecordTimeLabelColor(ratio: ratio)
        
        self.modeSelectView.setRatio(ratio)
        self.shutterView.setRatio(ratio)
        
        self.beautyView.setRatio(ratio)
        self.filterView.setRatio(ratio)
        self.contentView.setRatio(ratio)
        self.bulgeView.setRatio(ratio)
    }
    
    func setButtonsImage(ratio: ARGMediaRatio) {
        
        var beautyButtonImageName = self.kMainBottomFunctionBeautyButtonName
        var filterButtonImageName = self.kMainBottomFunctionFilterButtonName
        var contentButtonImageName = self.kMainBottomFunctionContentButtonName
        var bulgeButtonImageName = self.kMainBottomFunctionBulgeButtonName
        
        var imageColorString = ""
        if ratio == ._16x9 {
            imageColorString = "White"
        } else {
            imageColorString = "Black"
        }
        beautyButtonImageName.append(imageColorString)
        filterButtonImageName.append(imageColorString)
        contentButtonImageName.append(imageColorString)
        bulgeButtonImageName.append(imageColorString)

        beautyButtonImageName.append("_".appendLanguageCode())
        filterButtonImageName.append("_".appendLanguageCode())
        contentButtonImageName.append("_".appendLanguageCode())
        bulgeButtonImageName.append("_".appendLanguageCode())
        
        self.beautyButton.setImage(UIImage(named: beautyButtonImageName), for: .normal)
        self.beautyButton.setImage(UIImage(named: beautyButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)

        self.filterButton.setImage(UIImage(named: filterButtonImageName), for: .normal)
        self.filterButton.setImage(UIImage(named: filterButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)

        self.contentButton.setImage(UIImage(named: contentButtonImageName), for: .normal)
        self.contentButton.setImage(UIImage(named: contentButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)

        self.bulgeButton.setImage(UIImage(named: bulgeButtonImageName), for: .normal)
        self.bulgeButton.setImage(UIImage(named: bulgeButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)
    }
    
    func setRecordTimeLabelColor(ratio: ARGMediaRatio) {
        
        var textColor: UIColor = .clear
        var shadowColor: UIColor = .clear
        
        if ratio == ._16x9 {
            textColor = self.kMainBottomFunctionRecordTimeLabelRatioFullColor
            shadowColor = self.kMainBottomFunctionRecordTimeLabelShadowColor
        } else {
            textColor = self.kMainBottomFunctionRecordTimeLabelColor
        }

        self.recordTimeLabel.textColor = textColor
        
        self.recordTimeLabel.layer.shadowOffset = CGSize.zero
        self.recordTimeLabel.layer.shadowRadius = 1.0
        self.recordTimeLabel.layer.shadowOpacity = 0.5
        self.recordTimeLabel.layer.masksToBounds = false
        self.recordTimeLabel.layer.shouldRasterize = true
        self.recordTimeLabel.layer.shadowColor = shadowColor.cgColor
    }


    @IBAction func handleTapFrom(_ tap: UITapGestureRecognizer) {
        if tap.state == .ended {
            self.closeBottomFunctions()
        }
    }
    
    @IBAction func bottomButtonAction(_ sender: UIButton) {
        
        self.openBottomFunctions()
        
        let buttonType = MainBottomButtonType(rawValue: sender.tag)
        
        switch buttonType {
        case .Beauty:
            self.beautyView.open()
        case .Filter:
            self.filterView.open()
        case .Content:
            self.contentView.open()
        case .Bulge:
            self.bulgeView.open()
        default:
            break
        }
    }
    
    @IBAction func shutterButtonAction(_ sender: UIButton) {
        guard let delegate = self.delegate else { return }
        
        let mode = self.modeSelectView.mode
        switch mode {
        case .photo:
            delegate.photoButtonAction(sender)
        case .video:
            self.recordTimeLabel.isHidden = sender.tag.boolValue
            self.setModeSelectViewDisable(sender.tag.boolValue)
            delegate.videoButtonAction(sender)
        default:
            break
        }
    }
    
    private func setModeSelectViewDisable(_ disable: Bool) {
        self.modeSelectView.isUserInteractionEnabled = disable
    }

    func openBottomFunctions() {
        self.touchCancelView.isHidden = false
        
        self.isHidden = true
    }
    
    func closeBottomFunctions() {
        self.touchCancelView.isHidden = true
        
        self.beautyView.close()
        self.filterView.close()
        self.contentView.close()
        self.bulgeView.close()
        
        self.isHidden = false
    }
    
    func setRecordTime(_ time: Float) {
        let min: Int = Int(time / 60)
        let sec: Int = Int(time) - min * 60
        
        self.recordTimeLabel.text = String(format: "%02d:%02d", min, sec)
    }
}
